returnn/datasets/distrib_files.py

```diff
@@ -245,8 +267,32 @@ def init_seq_order(self, epoch: Optional[int] = None, seq_list=None, seq_order=N
         self._lazy_init_file_sizes()
         self._lazy_init_file_cache_proc()
 
-        # Cleanup and fill _files_order_cache.
         full_epoch_0idx = (epoch - 1) // self.partition_epoch
+
+        # Shard files across GPU workers
+        for k in list(self._files_for_this_worker.keys()):
+            if k < full_epoch_0idx:
+                del self._files_for_this_worker[k]
+        for ep_ in range(epoch, epoch + self.preload_next_n_sub_epochs + 1):
+            full_epoch_0idx_ = (ep_ - 1) // self.partition_epoch
+            if full_epoch_0idx_ in self._files_for_this_worker:
+                continue
+            if self.seq_ordering == "default":
+                files = self.files
+            elif self.seq_ordering == "random":
+                # do not use `self._get_random_seed_for_epoch` to exclude `random_seed_offset`
+                # to ensure an equal distribution in every worker
```

Comment:
This is wrong. We should use `_get_random_seed_for_epoch` in any case. We just should fix `_get_default_random_seed_offset` (just the same as for TF).

Reply:
I don't understand what you mean. If I used `_get_random_seed_for_epoch` here every horovod worker process would "mix in" its random_seed_offset and there would be different seeds per process for the initial sharding per worker process. Specifically, I don't understand what you mean with "fixing" `_get_default_random_seed_offset `. It seems like it does do exactly what it's supposed to do?

Reply:
> here every horovod worker process would "mix in" its random_seed_offset

No, this is only when the user configures `horovod_dataset_distribution = "random_seed_offset"`. But when the user configures `horovod_dataset_distribution = "shard"`, it will not do that. Despite, when the user configures an explicit `random_seed_offset`, it will also ignore the worker rank but just use the explicit `random_seed_offset`.

Currently for PyTorch, the `_get_default_random_seed_offset` does not do this. There is no such setting like `torch_distributed_dataset_distribution`, it always assumes "random_seed_offset". That's what I mean by fixing. `_get_default_random_seed_offset`/`_get_random_seed_for_epoch` should not depend on the worker rank when sharding is done. If they do, they don't work correct, i.e. should be fixed.